contracts/claimdrops.js

```diff
@@ -0,0 +1,346 @@
+      && ((!limit && list && Array.isArray(list))
+        || (!list && limit && typeof limit === 'string' && !api.BigNumber(limit).isNaN())), 'invalid params')) {
+    const token = await api.db.findOneInTable('tokens', 'tokens', { symbol });
+    const params = await api.db.findOne('params', {});
+
+    const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+    const timestamp = blockDate.getTime();
+    const expiryTimestamp = getTimestamp(expiry);
+    const maxExpiryTimestamp = api.BigNumber(timestamp)
+      .plus(params.maxExpiryTime).toNumber();
+
+    // get api.sender's utility and airdrop token balances
+    const utilityToken = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: UTILITY_TOKEN_SYMBOL });
+    const nativeToken = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol });
+
+    if (api.assert(token !== null, 'symbol does not exist')
+      // price checks
+      && api.assert(api.BigNumber(price).gt(0), 'price must be positive')
+      && api.assert(hasValidPrecision(price, HIVE_PEGGED_SYMBOL_PRECISION), 'price precision mismatch')
+      // pool checks
+      && api.assert(api.BigNumber(pool).gt(0), 'pool must be positive')
+      && api.assert(hasValidPrecision(pool, token.precision), 'pool precision mismatch')
+      // maxClaims check
+      && api.assert(api.BigNumber(maxClaims).gt(0), 'maxClaims must be positive number')
+      // expiry check
+      && api.assert(expiryTimestamp && expiryTimestamp > timestamp, 'invalid expiry')
+      && api.assert(expiryTimestamp <= maxExpiryTimestamp, 'expiry exceeds limit')
+      && api.assert(ownerType === 'user' || ownerType === 'contract', 'invalid ownerType')
+      && api.assert(await isValidowner(owner, ownerType), 'invalid owner')) {
+      const fee = api.BigNumber(params.feePerClaim).times(maxClaims)
+        .plus(params.creationFee)
+        .toFixed(UTILITY_TOKEN_PRECISION);
+
+      // balance checks
+      if (api.assert(utilityToken && utilityToken.balance
+        && api.BigNumber(utilityToken.balance).gte(fee), 'you must have enough tokens to cover the creation fee')
+        && api.assert(nativeToken && nativeToken.balance
+          && api.BigNumber(nativeToken.balance).gte(pool), 'you must have enough tokens to cover the claimdrop pool')) {
+        const claimdrop = {
+          claimdropId: api.transactionId,
+          symbol,
+          price,
+          remainingPool: pool,
+          remainingClaims: maxClaims,
+          claims: [],
+          expiry: expiryTimestamp,
+          owner,
+          ownerType,
+        };
+
+        // add list or limit to final claimdrop object
+        if (list) {
+          const parsedList = validateList(list, token.precision);
+          if (parsedList) {
+            claimdrop.list = parsedList;
+          } else return;
+        } else if (limit) {
+          if (api.assert(api.BigNumber(limit).gt(0), 'limit must be positive')
```

Comment:
Can you do a max size check for limit? Can be something like asserting that api.BigNumber(limit).lte(Number.MAX_SAFE_INTEGER) and also use the BigNumber isFinite function - https://mikemcl.github.io/bignumber.js/#isF